analyze.py

```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python
+import fileinput
+import os
+import re
+import sys
+import threading
+import time
+
+from server import Server
+from request import DuplicateEvent
+
+LOG_LOCATION = '/tmp/log_analyzer.out.{}'.format(os.getpid())
+LOG_FMT = ('Endpoint: {}\tTotal Requests: {}\t'
+           'Average Request Time: {}\tMedian Request Time: {}')
+
+def summarize_all(servers):
+    if not servers:
+        return
+    with open(LOG_LOCATION, 'w') as log:
+        for s in servers.itervalues():
+            log.write(LOG_FMT.format(*s.summarize())+'\n')
+    print 'Analysis complete, logs can be fount at {}'.format(LOG_LOCATION)
+
+def summarize_some(servers):
+    if not servers:
+        return
+    server_list = []
+    for s in servers.itervalues():
+        id, num_req, avg_req_time, med_req_time  = s.summarize()
+        server_list.append((id, num_req, avg_req_time, med_req_time))
+    busiest = sorted(server_list, cmp=lambda x,y: cmp(y[1], x[1]))[:6]
+    slowest = sorted(server_list, cmp=lambda x,y: cmp(y[2], x[2]))[:6]
+    print '====5 Busiest Endpoints===='
+    for b in busiest:
+        print LOG_FMT.format(*b)
+    print '====5 Slowest Endpoints===='
+    for s in slowest:
+        print LOG_FMT.format(*s)
+    print
+
+def main():
+    last_summary = time.time()
+    servers = dict()
+    try:
+        for line in fileinput.input():
+            if time.time() - last_summary >= 30.0:
+                summarize_some(servers)
+                last_summary = time.time()
+            try:
+                timestamp, endpoint, req_id, event = line.split(' ')
+            except ValueError:
+                sys.stderr.write('Invalid log format: {}'.format(line))
+                continue
+            if endpoint not in servers:
+                servers[endpoint] = Server(endpoint)
```

Comment:
It looks like `servers` grows without bound along with the number of endpoints your program has ever seen.  This would break the "finite memory footprint" constraint.